File: two_sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        int[] answer = new int[2];
        
        //First, all of the array's element is put to a map
        for(int i=0;i<nums.length;i++){
            map.put(nums[i],i);
        }
        //Find target with containsKey using : target = nums[i] + nums[j] -> target-nums[i] = nums[j]
        for(int i=0;i<nums.length;i++){
            int find = target - nums[i];
            
            if(map.containsKey(find) && map.get(find) != i){ //Prevent using same index value twice
                answer[0] = i;
                answer[1] = map.get(find); //find will be a key
            }
        }
        return answer;
    }
}

File: TwoSum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        
        num_dict = dict()   #map
        for i, num in enumerate(nums):
            num_dict[num] = i   #Numbers become key, and indeces become value
        
        for i,num in enumerate(nums):
            complement = target - num 
            if complement in num_dict and i != num_dict[complement]: #find complement as a key in the map, and not the same index
                return [i, num_dict[complement]]

 
#One for loop but same time complexity

class Solution:
    def twoSum(self, nums:List[int], target: int) -> List[int]:
        
        num_dict = dict() 
        
        for i, num in enumerate(nums):  #Iteration nums list
            complement = target - num 
            if complement in num_dict:  #Find complement in num_dict by target - nums[i] = nums[i2] 
                return [i, num_dict[complement]] #if found, return index and value of the key 
            num_dict[nums] = i #If the complement does not exist, then add it to num_dict
